### ISSUE TITLE
List index out of range when using tokenizer.learn_bpe

### ISSUE THREAD

**Issue by VladPetk** [Natooz/MidiTok#90]
Hi!

Here's the relevant part of my code:
```
# Tokenizer Parameters
TOKENIZER_PARAMS = {
    "pitch_range": (21, 109),
    "nb_velocities": 32,
    "special_tokens": [],
    "use_chords": True,
    "use_rests": False,
    "use_tempos": False,
    "use_time_signatures": False,
    "use_programs": False,
}
config = miditok.TokenizerConfig(**TOKENIZER_PARAMS)

# Creates the tokenizer
tokenizer = miditok.Structured(config)

# tokenize the dataset
tokenizer.tokenize_midi_dataset([x[0] for x in all_comps], "m_a_tokens_no_bpe")
# Results in 1498 tokenized Midi files

#Create a list of paths to tokens
token_list = []
for (dirpath, dirnames, filenames) in os.walk("D:\\musicai\\raw_data\\processed\\m_a_tokens_no_bpe"):
    token_list += [os.path.join(dirpath, file) for file in filenames]

#Perform bpe learning
tokenizer.learn_bpe(
    vocab_size=1000,
    tokens_paths=token_list,  out_dir="D:\\musicai\\raw_data\\processed\\m_a_tokens_no_bpe_aug",
)
```
The `tokenizer.learn_bpe` successfully loads the token files, but then I get an error "_list index out of range_"  in this line of code: `if isinstance(tokens[0], (list, tuple)):` which is part of the `_tokens_to_ids` function.

I tried perfroming `_tokens_to_ids` on a single set of tokens (tokens from one MIDI file) by first loading its JSON file and get the same error:
```
with open(token_list[1], "r") as json_file:
    example = json.load(json_file)
tokenizer._tokens_to_ids(example)
```

However, if I instead point to the actual location of IDs in the JSON file, it works as expected: `tokenizer._tokens_to_ids(example["ids"][0])` . This outputs a list of tokens (shouldn't the function actually do the opposite - convert tokens to IDs?)

Am I doing something wrong or is this a bug?

I also tried simply saving only the `["ids"][0]` part of every file and passing that list to `tokenizer.learn_bpe`, but that doesn't work either.

Any help would be appreciated!

**Comment by VladPetk** [comment]
And I think I found the reason: I didn't have any special tokens in my tokenizer and this part of the code seems to require at least one to not throw an error: `special_tokens_bytes = self._ids_to_bytes(self._tokens_to_ids([f"{tok}_None" for tok in self.config.special_tokens])`

Is there a way to do this without special tokens? The only workaround I see is tokenizing the data without the special tokens, then creating a new tokenizer with the same parameters except for an extra special token and using the the tokenizer to learn bpe.

**Comment by Natooz** [comment]
Hello,

I'm looking into it, I didn't run the code but I think the issue comes from the fact that the list of `special_tokens` is empty. Could you try to run it by putting something?

`_tokens_to_ids` actually calls the `MIDITokenizer.__get__()` magic method, which automatically detects if the item is an id (int) or token (str). The token is the string form (e.g. `Pitch_53`) and an id a unique integer id associated to a token. So what you are getting is "normal", in the way the method are working, the method shouldn't receive ids. :) But indeed the code could be optimized by directly accessing the token -> id dictionary, that would also raise an error in case the method is misused.

**Comment by Natooz** [comment]
You got faster than me 😄

I'll make a fix to make the method work without special tokens!
On another note however, you might probably want to add a `PAD` token in the special tokens, unless you do not plan to train and use your model with a batch size >1, or with sequences of the exact same size without padding.

**Comment by VladPetk** [comment]
Thanks for the quick reply! 

Makes sense to add a `PAD` of course. It's just that the way my dataset creator and loader are setup now, the `PAD` tokens are added directly there (since I'm taking subsequences of different length dependinng on some conditions). But I suppose I could pass the same token as is created in the tokenizer.

**Comment by Natooz** [comment]
I don't know your `PAD` tokens are added, but if you do not "register" a padding token in the tokenizer and do not use the same associated id (often 0), chances are you will be padding with another "real" token, which could be problematic.

Usually, the PyTorch datasets / data loaders / collators receive the padding token of the tokenizer when they are created. (this also the case with text)
